calico/reference/resources/felixconfig.md

```diff
@@ -150,12 +150,16 @@ The `RouteTanbleRange` option is now deprecated in favour of [RouteTableRanges](
 | min      | Minimum index to use | int    |
 | max      | Maximum index to use | int    |
 
-Each `RouteTableRange` designates a set of routing tables available to Calico. Modern Linux kernels can theorectically support more than the documented 252 additional tables. However, allocating more than the default number of tables is not as simple as increasing the max table index, since the increased range would encompass the reserved tables, and could lead to Calico clashing with the kernel.
+Each item in the `RouteTableRanges` list designates a range of routing tables available to Calico. By default, Calico will use a single range of `1-10000`, and will automatically ignore Linux's default reserved tables `253-255`. However, it's possible that other table ranges may also be reserved by third-party systems unknown to Calico. In that case, multiple ranges can be defined to target tables below and above the sensitive ranges:
+```sh
+# target tables 65-99, and 256-1000, skipping 100-255
+calicoctl patch felixconfig default --type=merge -p '{"spec":{"routeTableRanges": [{Min: 65, Max: 99}, {Min: 256, Max: 10000}] }}
+```
+
 
-Instead, multiple ranges can be defined to designate tables below and above the reserved ranges:
-`calicoctl patch felixconfig default --type=merge -p '{"spec":{"routeTableRanges": [{Min: 1, Max: 250}, {Min: 256, Max: 1000}] }}`
+*Note*, for performance reasons, the maximum total number of routing tables that Felix will accept is 65535 (or 2*16).
 
-If `RouteTableRange` and `RouteTableRanges` are both present, the latter will take precedence.
+If the deprecated `RouteTableRange` option is explicitly set, it will be honoured in favour of `RouteTableRanges`.
```

Comment:
I think I'd have jumped the other way if both are set.  If both are explicitly set, I'd use the value from RouteTableRanges.

Reply:
I initially wanted to go that route, but I encountered a problem:
 - remove `RouteTableRange` default so we can check when it's empty
 - `RouteTableRanges` still has a non-zero default, so it will always look "set"-ish
 - if the user the deprecated arg and left the new one as default, while I _could_ check that `RouteTableRanges` arg isnt _default_, and go with the deprecated arg, there would be ambiguity - if they for some reason explicitly set `RouteTableRanges` to the default val, I wouldn't be able to tell that apart :thinking: 
 
I didn't immediately see a way from within the config function to differentiate between set (to the default) and unset default, so I opted to give the old param precedence when it's there, since if it's unset, it'll definitely be a zero-val

Hope that makes sense. maybe Im missing a way around this?

Reply:
One option would be to remove both their defaults in the Felix config_params.go and then do the defaulting "manually" in the getter function.  We do it that way to handle a couple of other "more complicated" defaults.

Reply:
Thats a great suggestion :raised_hands: will implement.

Reply:
I believe the way that Kubernetes has handled API changes from singular to plural is through validation that both are set consistently.

e.g., if `RouteTableRange` and `RouteTableRanges` are set, they need to be consistent with each other else they will error. 

That might just mean "Return an error on the API if both are set" - that seems like the cleanest behavior to me.

Reply:
The way I think about this is that the code shouldn't be assuming or trying to guess the users intent - if the user gives us two pieces of conflicting intent via the API, we need to tell the user "Hey, you're not making sense".

Reply:
I like the suggestion, personally am happy with either approach. Are there other examples of either solution being done before in our code? If this is the first occurence, I'm happy to put the best foot forward, but Im also conscious of getting choice paralysis here as I'd be rewriting this piece of the PR for the third time :laughing:  if it's something that we're ambiguous about, I might acknowledge the suggestion but leave it off this time around. WDYT?

Reply:
I think the issue with FelixConfiguration that makes my suggestion above not entirely viable is that the config that Felix ultimately sees might have both of these set, even if we prevent them from both being set on a single API object, since Felix merges config from multiple API objects as well as env vars, files, etc. So we need to think of this in two ways: What does the API do with this, and what does Felix do with this.

I think we should do effectively what Shaun suggested above, but with a little bit of API validation so that on a single object only one of these is ever set. i.e.,

- The API validation code should prevent both of these from being set on the same object (libcalico-go validation func) 
- Remove Felix config_param defaulting per @fasaxc's suggestion above. 
- Felix code should emit a warning log if it sees both config options as being set.
- Felix code should emit a warning log if it sees the old, deprecated option as being set.
- Felix code prefers the new option if both are set.

I think that means you probably don't need to re-work this logic again.

Reply:
Sounds reasonable! Will implement :smiley: thanks guys